### PR TITLE
Setup `HrmpChannelAccepted` handler with `pallet_xcm` for initiating XCM version discovery when openning HRMP channel

### PR DESCRIPTION
**Pull request by bkontur**
Closes: https://github.com/paritytech/polkadot-sdk/issues/922
Implements: https://github.com/paritytech/polkadot-sdk/issues/3692 (see: `Proposed solution part 2 - initiate VersionDiscoveryQueue when HRMP is accepted`).

When we utilize a parachain's `pallet_xcm` extrinsics or other functionality that employs `XcmpQueue` as an `XcmRouter`, there is a possibility of encountering a situation where it is the first message to the desired destination, and we do not know a supported XCM version of the destination. In such cases, `pallet_xcm` includes a feature called "version discovery," which is initiated. This is a one-time operation. Therefore, this fix aims to initiate "version discovery" as soon as possible when an HRMP channel is created, with best effort.

In other words, when parachains establish an HRMP channel, we immediately initiate XCM version discovery and don't wait for the very first user request.

This PR:
- introduces `AllowHrmpNotificationsFromRelayChain` barrier for allowing HRMP notifications just from the relay chain (to fulfill safety assumptions - [see](https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/xcm/src/v4/mod.rs#L532))
- adds the implementation of `HandleHrmpChannelAccepted` for `pallet_xcm`, so `pallet_xcm` instance can be used for `xcm_executor::Config::HrmpChannelAcceptedHandler`(https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/xcm/xcm-executor/src/config.rs#L121-L122)
- sets `HandleHrmpChannelAccepted` for all cumulus/testnets

### CODE REVIEW

File: polkadot/xcm/xcm-builder/src/tests/barriers.rs [hunk 1]
```diff
@@ -315,56 +315,150 @@ fn allow_subscriptions_from_should_work() {
...
+		Parachain(1).into_location(),
+		Err(ProcessMessageError::Unsupported),
+	);
+	assert_should_execute(
+		vec![UnsubscribeVersion],
+		Parachain(1).into_location(),
+		Err(ProcessMessageError::Unsupported),
+	);
 
-	let test_data = vec![
```
[the remark this answers is not in the document]
Comment (serban300) [reply polkadot/xcm/xcm-builder/src/tests/barriers.rs:332]: Thanks ! Will take a look.

File: polkadot/xcm/pallet-xcm/src/lib.rs [hunk 2]
```diff
@@ -3049,6 +3050,35 @@ impl<T: Config> CheckSuspension for Pallet<T> {
...
 }
 
+/// The implementation of `HandleHrmpChannelAccepted` which can initiate version discovery (if
+/// needed).
+///
+/// Note: This is useful for parachains when establishing HRMP channels with other parachains.
+impl<T: Config> HandleHrmpChannelAccepted for Pallet<T> {
+	fn handle(recipient: u32) -> XcmResult {
+		// how would the actual runtime see the `recipient` as `Parachain(recipient)`
+		let dest = match T::UniversalLocation::get().global_consensus() {
```
Comment (svyatonik) [inline polkadot/xcm/pallet-xcm/src/lib.rs:3060]: Just curious - I though that `UniversalLocation` is assumed to be started with a `GlobalConsensus`. Then why would we need that match here? A defensive programming or something else?

**Review by svyatonik — APPROVED** [review]

File: polkadot/xcm/xcm-builder/src/tests/barriers.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (serban300) [reply polkadot/xcm/xcm-builder/src/tests/barriers.rs:332]: Looks good ! Thank you !

**Review by serban300 — APPROVED** [review]

File: polkadot/xcm/pallet-xcm/src/lib.rs (hunk 2 above)
Comment (bkontur) [reply polkadot/xcm/pallet-xcm/src/lib.rs:3060]: > Just curious - I though that `UniversalLocation` is assumed to be started with a `GlobalConsensus`. Then why would we need that match here? A defensive programming or something else?

Yes, good point, exactly - defensive and generic.

Originally, I wanted to use just `let dest = ParentThen(Parachain(recipient).into()).into();`, but yeah there is a `Parent` which is also not that generic.

Yes, there is that thing with `GlobalConsensus` you mentioned, which I guess it's kind of unwritten rule or assumption that runtime's UniversalLocation starts with `GlobalConsensus` - I don't remember exactly but without `GlobalConsensus` something fails in `XcmExecutor` (maybe just `ExportMessage` or something with `WithComputedOrigin`...)

The Hrmp notifications have just `paraId: u32` and we need to convert it to `xcm::Location`, so I was thinking this way:
`*Parachain(recipient)*.relative_to(&T::UniversalLocation::get());`.


```
        #[test]
	fn test() {
		// E.g. we are on AssetHub
		frame_support::parameter_types! {
			pub UniversalLocation: InteriorLocation = [GlobalConsensus(Rococo), Parachain(1000)].into();
		}

		// We opened channel with 1003
		let recipient = 1003;

		// convert 1003 from the point of view of `UniversalLocation`
		let dest = match UniversalLocation::get().global_consensus() {
			Ok(global_consensus) =>
				Junctions::from([GlobalConsensus(global_consensus), Parachain(recipient)]),
			Err(_) => Junctions::from([Parachain(recipient)]),
		}.relative_to(&UniversalLocation::get());

		// expected result, how AssetHub sees sibling parachain
		assert_eq!(dest, Location { parents: 1, interior: Junctions::X1([Parachain(1003)].into()) });
	}
```

And when I was checking other runtimes, I saw that we have some minimal template without `GlobalConsensus`: https://github.com/paritytech/polkadot-sdk/blob/master/templates/parachain/runtime/src/configs/xcm_config.rs#L29, so thats why I ended up with that match.

@svyatonik Do you have any other idea how to handle a conversion from `paraId: u32` to the `Location` with `T::UniversalLocation`? I was thinking about some adapter with which would wrap `Convert<u32, Location>`, `Self::get_version_for` with `Self::note_unknown_version`, but it could just complicate things and not sure if it is worth.

Comment (acatangiu) [reply polkadot/xcm/pallet-xcm/src/lib.rs:3060]: > I though that UniversalLocation is assumed to be started with a GlobalConsensus.

Yes, this is ideally where we should get, but not all parachains have done this yet..

Everything except bridging works fine without it too, so there is also no big incentive to do it.

Comment (acatangiu) [reply polkadot/xcm/pallet-xcm/src/lib.rs:3060]: > And when I was checking other runtimes, I saw that we have some minimal template without GlobalConsensus: https://github.com/paritytech/polkadot-sdk/blob/master/templates/parachain/runtime/src/configs/xcm_config.rs#L29

We should sanitize whole repo and make sure all our runtimes test, mock, etc have the `GlobalConsensus` too so we create a consistent best-practice.

Comment (svyatonik) [reply polkadot/xcm/pallet-xcm/src/lib.rs:3060]: > Originally, I wanted to use just let dest = ParentThen(Parachain(recipient).into()).into();, but yeah there is a Parent which is also not that generic.

But OTOH we know that this notification comes from the relay chain (because it notifies us about new HRMP channels), right? So maybe `Parent` would be enough here :) Not sure if it'll be changed

> And when I was checking other runtimes, I saw that we have some minimal template without GlobalConsensus: https://github.com/paritytech/polkadot-sdk/blob/master/templates/parachain/runtime/src/configs/xcm_config.rs#L29, so thats why I ended up with that match.

Maybe we should fix that in this case. E.g. some bridging adapters (that's where I got this assumption - e.g. `UnpaidRemoteExporter`) are assuming that `UniversalLocation` starts with `GlobalConsensus`. But that's not related to this PR ofc.

Thank you for the explanation - I've just been curios about that && I wasn't planning to request changes or smth like that :)

**Review by acatangiu — APPROVED** [review]

**Comment by acatangiu** [comment]
@bkontur is this good to merge?

**Comment by bkontur** [comment]
> @bkontur is this good to merge?

not yet, I want to check comments and add also closing part, where we clean storage for unneded stuff

File: polkadot/xcm/pallet-xcm/src/lib.rs (hunk 2 above)
Comment (bkontur) [reply polkadot/xcm/pallet-xcm/src/lib.rs:3180]: > > And when I was checking other runtimes, I saw that we have some minimal template without GlobalConsensus: https://github.com/paritytech/polkadot-sdk/blob/master/templates/parachain/runtime/src/configs/xcm_config.rs#L29
> 
> We should sanitize whole repo and make sure all our runtimes test, mock, etc have the `GlobalConsensus` too so we create a consistent best-practice.

Good point, attempt to sanitize (where possible): https://github.com/paritytech/polkadot-sdk/pull/4025/commits/9c5a91893a4fe37561bba3140e6fea4a4a858391

**Comment by bkontur** [comment]
- [ ] investigating one more test if this is enough and aligned with `process_hrmp_open_channel_requests` / `initializer_on_new_session`

**Comment by bkontur** [comment]
Change to draft, because `HrmpNotificationAccepted` won't work and need to find another solution: https://github.com/paritytech/polkadot-sdk/issues/922#issuecomment-2058868951

Opened another PR: https://github.com/paritytech/polkadot-sdk/pull/4156 which contains still relevant changes but with reverted `HrmpNotificationAccepted` handler stuff.
